File: old_contents/boj_codes/Python3/2252.py
# https://www.acmicpc.net/problem/2252
# 20201105

import sys


class Node:
    def __init__(self, item):
        self.item = item
        self.indegree = 0
        self.connected = []


class DAG:
    def __init__(self, N_students):
        self.nodes = [Node(0)]
        self.len = N_students
        self.visited = [False] * (N_students + 1)

        for i in range(1, N_students + 1):
            self.nodes.append(Node(i))

    def add_edge(self, idx1, idx2):
        self.nodes[idx1].connected.append(idx2)
        self.nodes[idx2].indegree += 1

    def dfs(self, idx, result_list):
        self.visited[idx] = True
        for sub_idx in self.nodes[idx].connected:
            if not self.visited[sub_idx]:
                self.dfs(sub_idx, result_list)

        result_list.append(idx)
        return

    def topological_sort(self):
        result_list = []

        for idx in range(1, self.len + 1):
            if not self.visited[idx]:
                self.dfs(idx, result_list)

        result_list.reverse()
        return result_list


if __name__ == "__main__":
    N, M = map(int, sys.stdin.readline().split())
    dag = DAG(N)
    for _ in range(M):
        idx1, idx2 = map(int, sys.stdin.readline().split())
        dag.add_edge(idx1, idx2)

    result = dag.topological_sort()
    for res in result:
        print(res, end = ' ')

File: old_contents/boj_codes/C++17/2252.cpp
// https://www.acmicpc.net/problem/2252
// First Written : 20210105
// Last Modified : 20210105

#include <cstdio>
#include <vector>

using namespace std;

struct Edge{
    int start;
    int end;
};

class DAG{
private:
    int size;
    bool* visitied;
    vector<vector<int>> nodes;

public:
    DAG(int n){
        nodes.resize(n + 1);
        visitied = new bool[n + 1]();
        this->size = n;
    }

    ~DAG(){
        delete [] visitied;
    }
    void add_edge(Edge e){
        nodes[e.end].push_back(e.start);
    }

    void topological_sort(vector<int>& result){
        for(int i = 1; i <= size; i++)
            if(!visitied[i])
                dfs(i, result);
    }

    void dfs(int idx, vector<int>& result){
        visitied[idx] = true;
        for(auto sub_idx : nodes[idx])
            if(!visitied[sub_idx])
                dfs(sub_idx, result);

        result.push_back(idx);
    }
};

int main(){
    int N, M;
    scanf("%d %d", &N, &M);

    Edge e;
    DAG graph = DAG(N);
    vector<int> result;

    while(M--){
        scanf("%d %d", &e.start, &e.end);
        graph.add_edge(e);
    }

    graph.topological_sort(result);
    for(auto it: result){
        printf("%d ", it);
    }

    return 0;
}